File: packages/mcp-server/src/tools/agent-runs.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';
import { pmApi, PmApiError } from '../pm-api.js';

function ok(data: unknown) {
  return { content: [{ type: 'text' as const, text: JSON.stringify(data) }] };
}

function err(message: string) {
  return { content: [{ type: 'text' as const, text: `Error: ${message}` }] };
}

export function registerAgentRunsTools(server: McpServer) {
  // POST /api/agent-runs
  server.tool(
    'emit_event',
    'Emit an agent run event (creates a new agent_runs record)',
    {
      agent_id: z.string().describe('Team member ID of the agent'),
      trigger: z.string().describe('Event trigger name'),
      model: z.string().optional().describe('Model used'),
      story_id: z.string().optional().describe('Related story ID'),
      memo_id: z.string().optional().describe('Related memo ID'),
      result_summary: z.string().optional().describe('Summary of the run result'),
      status: z.enum(['running', 'completed', 'failed']).optional().describe('Run status (default: completed)'),
      error_message: z.string().optional().describe('Error message if failed'),
      input_tokens: z.number().optional().describe('Input token count'),
      output_tokens: z.number().optional().describe('Output token count'),
      started_at: z.string().optional().describe('ISO timestamp when run started'),
      finished_at: z.string().optional().describe('ISO timestamp when run finished'),
    },
    async ({ agent_id, trigger, model, story_id, memo_id, result_summary, status, error_message, input_tokens, output_tokens, started_at, finished_at }) => {
      try {
        const body: Record<string, unknown> = { agent_id, trigger };
        if (model !== undefined) body.model = model;
        if (story_id !== undefined) body.story_id = story_id;
        if (memo_id !== undefined) body.memo_id = memo_id;
        if (result_summary !== undefined) body.result_summary = result_summary;
        if (status !== undefined) body.status = status;
        if (error_message !== undefined) body.error_message = error_message;
        if (input_tokens !== undefined) body.input_tokens = input_tokens;
        if (output_tokens !== undefined) body.output_tokens = output_tokens;
        if (started_at !== undefined) body.started_at = started_at;
        if (finished_at !== undefined) body.finished_at = finished_at;
        const data = await pmApi('/api/agent-runs', { method: 'POST', body: JSON.stringify(body) });
        return ok(data);
      } catch (e) {
        return err(e instanceof PmApiError ? e.message : String(e));
      }
    },
  );

  // PATCH /api/agent-runs/[id]
  server.tool(
    'update_run_status',
    'Update an existing agent run status (supports retry chain)',
    {
      run_id: z.string().describe('Agent run ID to update'),
      status: z.enum(['running', 'completed', 'failed']).describe('New status'),
      error_message: z.string().optional().describe('Error message if failed'),
      result_summary: z.string().optional().describe('Result summary'),
      input_tokens: z.number().optional().describe('Input token count'),
      output_tokens: z.number().optional().describe('Output token count'),
      cost_usd: z.number().optional().describe('Cost in USD'),
      started_at: z.string().optional().describe('ISO timestamp when run started'),
      finished_at: z.string().optional().describe('ISO timestamp when run finished'),
    },
    async ({ run_id, status, error_message, result_summary, input_tokens, output_tokens, cost_usd, started_at, finished_at }) => {
      try {
        const body: Record<string, unknown> = { status };
        if (error_message !== undefined) body.error_message = error_message;
        if (result_summary !== undefined) body.result_summary = result_summary;
        if (input_tokens !== undefined) body.input_tokens = input_tokens;
        if (output_tokens !== undefined) body.output_tokens = output_tokens;
        if (cost_usd !== undefined) body.cost_usd = cost_usd;
        if (started_at !== undefined) body.started_at = started_at;
        if (finished_at !== undefined) body.finished_at = finished_at;
        const data = await pmApi(`/api/agent-runs/${run_id}`, { method: 'PATCH', body: JSON.stringify(body) });
        return ok(data);
      } catch (e) {
        return err(e instanceof PmApiError ? e.message : String(e));
      }
    },
  );

  // GET /api/agent-runs?project_id=X&limit=N
  server.tool(
    'poll_events',
    'Poll recent agent run events for a project',
    {
      project_id: z.string().describe('Project ID'),
      limit: z.number().optional().describe('Max results (default: 20)'),
    },
    async ({ project_id, limit }) => {
      try {
        const params = new URLSearchParams({ project_id });
        if (limit !== undefined) params.set('limit', String(limit));
        const data = await pmApi(`/api/agent-runs?${params.toString()}`);
        return ok(data);
      } catch (e) {
        return err(e instanceof PmApiError ? e.message : String(e));
      }
    },
  );
}

File: backend/sprintable_mcp/tools/agent_runs.py
"""에이전트 런 MCP 도구 (3개)."""
from __future__ import annotations

from typing import Literal

from mcp.types import TextContent

from ..api_client import client
from ..response import err, ok
from ..schemas import SprintableInput

RunStatus = Literal["running", "completed", "failed"]


class EmitEventInput(SprintableInput):
    agent_id: str
    trigger: str
    model: str | None = None
    story_id: str | None = None
    memo_id: str | None = None
    result_summary: str | None = None
    status: RunStatus | None = None
    error_message: str | None = None
    input_tokens: int | None = None
    output_tokens: int | None = None
    started_at: str | None = None
    finished_at: str | None = None


class UpdateRunStatusInput(SprintableInput):
    run_id: str
    status: RunStatus
    error_message: str | None = None
    result_summary: str | None = None
    input_tokens: int | None = None
    output_tokens: int | None = None
    cost_usd: float | None = None
    started_at: str | None = None
    finished_at: str | None = None


class PollEventsInput(SprintableInput):
    recipient_id: str | None = None
    event_type: str | None = None
    limit: int | None = None
    cursor: str | None = None  # 이전 호출의 X-Next-Cursor 헤더 값을 그대로 넘기면 다음 페이지.


async def emit_event(args: EmitEventInput) -> list[TextContent]:
    """에이전트 런 이벤트 발행."""
    try:
        body: dict = {
            "agent_id": args.agent_id, "trigger": args.trigger,
            "project_id": client.require_project_id(),  # E-MCP-OPT ff6cb90d: try 안에서 호출(가이드 에러 캡처).
        }
        for field in ("model", "story_id", "memo_id", "result_summary", "status",
                      "error_message", "input_tokens", "output_tokens", "started_at", "finished_at"):
            val = getattr(args, field)
            if val is not None:
                body[field] = val
        return ok(await client.post("/api/v2/agent-runs", json=body))
    except Exception as exc:
        return err(str(exc))


async def update_run_status(args: UpdateRunStatusInput) -> list[TextContent]:
    """에이전트 런 상태 업데이트."""
    body: dict = {"status": args.status}
    for field in ("error_message", "result_summary", "input_tokens", "output_tokens",
                  "cost_usd", "started_at", "finished_at"):
        val = getattr(args, field)
        if val is not None:
            body[field] = val
    try:
        return ok(await client.patch(f"/api/v2/agent-runs/{args.run_id}", json=body))
    except Exception as exc:
        return err(str(exc))


async def poll_events(args: PollEventsInput) -> list[TextContent]:
    """에이전트 수신 대기 이벤트 폴링. 더 있으면(has_more) 다음 페이지 안내가 별도 텍스트
    블록으로 붙는다 — 단 이 도구는 순수 read(이 MCP 계층엔 mark_delivered를 부르는 도구가
    없음)라 다른 list_* 도구와 달리 cursor 없이 그냥 다시 부르면 같은 상위 N건이 그대로
    다시 온다(«비워지지» 않음). 안내 문구가 그 함정을 명시한다."""
    from .stories import _has_more_from_headers

    recipient = args.recipient_id or client.member_id
    params: dict = {"recipient_id": recipient}
    if args.event_type:
        params["event_type"] = args.event_type
    if args.limit:
        params["limit"] = args.limit
    if args.cursor:
        params["cursor"] = args.cursor
    try:
        items, headers = await client.get_with_headers("/api/v2/events/pending", params=params)
        has_more, next_cursor = _has_more_from_headers(headers, items)
        blocks = ok(items)
        if has_more:
            cursor_hint = f'cursor="{next_cursor}"' if next_cursor else "cursor(서버가 next_cursor를 안 줌 — 호출부 확인 필요)"
            blocks.append(TextContent(
                type="text",
                text=(
                    f"※ 더 있음 — 이 응답은 {len(items)}건까지만 포함(전량 아님). "
                    f"이 도구는 mark_delivered를 부르지 않는 순수 read라 cursor 없이 다시 불러도 같은 상위 "
                    f"{len(items)}건이 그대로 다시 온다(pending이 비워지지 않음) — 나머지를 보려면 "
                    f"poll_events를 {cursor_hint}로 다시 호출."
                ),
            ))
        return blocks
    except Exception as exc:
        return err(str(exc))
